### PR TITLE
Feature/toolboxes refactoring

### PR DESCRIPTION
**Pull request by vincentchabannes**
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes?
- [x] Have you successfully run the Feel++ testsuite with your changes locally?

-----

Major update of the toolboxes libraries :
 - change json format : 
    - add Models section  (remove the Model key)
    - Materials section : markers used in a materials can be specified ( the name is now key of a material)
    - PostProcess sections :  (i) add Exports section ( no Fields) and  (ii) can apply several postprocess with respect to a subphysic in  a multiphysics toolbox 
 - directory, where results are exported, are define now by --directory options (replace exporter.directory option)
 - rename thermodyn -> heattransfer
 - refactoring of solidmechanics toolboxes : only one class (no base)
 - add electric toolbox + refactoring of thermoelectric
 - major clean of cmake macro for the toolbox lib generation
 - thermal/electric conductivity can be an expression which depends on temperature
 - add heatfluid toolbox (remove dependencies in fluidmec)
 - anisotropic heat conductivity (only implemented in linear)

### CODE REVIEW

**Comment by vincentchabannes** [comment]
@prudhomm the merge is ready. Have you time to review

**Comment by prudhomm** [comment]
I will do that over the weekend

File: feel/feelcore/environment.hpp [hunk 1]
```diff
@@ -611,7 +611,11 @@ class FEELPP_EXPORT Environment : boost::noncopyable
     //! the application directory, application results are stored there
     //! the directory is controlled by changeRepository
     static std::string appRepository();
-    
+
+    //! the application directory without np, application results are stored there
+    //! the directory is controlled by changeRepository
+    static std::string appRepositoryWithoutNumProc();
```
Comment (prudhomm) [inline feel/feelcore/environment.hpp]: why this change? what is the rationale?
is it to reduce the number of directories?
we have too many of them anyway. We need to radical change in the data handling

File: feel/feelmodels/modelexpression.hpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+ */
+#ifndef FEELPP_MODELEXPRESSION_HPP
+#define FEELPP_MODELEXPRESSION_HPP 1
+
+#include <feel/feelvf/expr.hpp>
+#include <feel/feelvf/ginac.hpp>
+
+namespace Feel {
+
+class FEELPP_EXPORT ModelExpression
```
Comment (prudhomm) [inline feel/feelmodels/modelexpression.hpp]: the class is exported so please add some doxygen like documentation here to explain what the class is doing

File: feel/feelmodels/modelexpression.hpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+private :
+    std::map<int,double> M_values;
+    boost::optional<expr_scalar_type> M_exprScalar;
+    boost::optional<expr_vectorial2_type> M_exprVectorial2;
+    boost::optional<expr_vectorial3_type> M_exprVectorial3;
+    boost::optional<expr_matrix22_type> M_exprMatrix22;
+    boost::optional<expr_matrix33_type> M_exprMatrix33;
+};
+
+class FEELPP_EXPORT ModelExpressionScalar : private ModelExpression
```
Comment (prudhomm) [inline feel/feelmodels/modelexpression.hpp]: add doxygen documentation
use Feel++ format for that

File: feel/feelmodels/modelmaterials.hpp [hunk 4]
```diff
@@ -460,7 +485,7 @@ material( ModelMaterials::value_type const& m )
 }
 
 FEELPP_EXPORT inline std::string
-marker( ModelMaterials::value_type const& m )
+name( ModelMaterials::value_type const& m )
```
Comment (prudhomm) [inline feel/feelmodels/modelmaterials.hpp]: won't it break some code to rename such a function?

Comment (prudhomm) [reply feel/feelmodels/modelmaterials.hpp]: I think you should duplicate the function, rename it and `FEELPP_DEPRECATED` `marker`  if you want to do that

File: feel/feelmodels/modelmodels.hpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+#define FEELPP_MODELMODELS_HPP 1
+
+#include <boost/property_tree/ptree.hpp>
+#include <feel/feelcore/feel.hpp>
+
+namespace Feel {
+
+namespace pt = boost::property_tree;
+
+class FEELPP_EXPORT ModelModel
```
Comment (prudhomm) [inline feel/feelmodels/modelmodels.hpp]: add documentation to the class for Doxygen and understand what it is doing. 
The name seems quite a bit awkward

File: feel/feelmodels/modelmodels.hpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    ModelModel( ModelModel const& ) = default;
+    ModelModel( ModelModel && ) = default;
+    pt::ptree const& ptree() const { return M_ptree; }
+    std::string const& equations() const { return M_equations; }
+  private:
+    pt::ptree M_ptree;
+    std::string M_equations;
+};
+
+class FEELPP_EXPORT ModelModels : public std::map<std::string,ModelModel>
```
Comment (prudhomm) [inline feel/feelmodels/modelmodels.hpp]: add documentation

File: feel/feelmodels/modelpostprocess.cpp [hunk 7]
```diff
@@ -250,20 +302,19 @@ ModelPostprocess::setup()
...
             }
         }
     }
+}
 
 
-}
 
 void
 ModelPostprocess::saveMD(std::ostream &os)
```
Comment (prudhomm) [inline feel/feelmodels/modelpostprocess.cpp]: this function should be marked as deprecated

Comment (prudhomm) [reply feel/feelmodels/modelpostprocess.cpp]: this will be superseded by #939

File: feel/feelmodels/modelpostprocess.hpp [hunk 8]
```diff
@@ -35,6 +35,28 @@ namespace Feel {
 
 namespace pt =  boost::property_tree;
 
+class FEELPP_EXPORT ModelPostprocessExports
```
Comment (prudhomm) [inline feel/feelmodels/modelpostprocess.hpp]: add Doxygen documentation

File: feel/feelmodels/modelpostprocess.hpp [hunk 9]
```diff
@@ -182,7 +204,7 @@ public :
     std::string M_directoryLibExpr;
 };
 
-class FEELPP_EXPORT ModelPostprocess: public std::map<std::string,std::vector<std::string>>
+class FEELPP_EXPORT ModelPostprocess//: public std::map<std::string,std::vector<std::string>>
```
Comment (prudhomm) [inline feel/feelmodels/modelpostprocess.hpp]: Add Doxygen Documentation

Comment (prudhomm) [reply feel/feelmodels/modelpostprocess.hpp]: remove the comment so that it is not misleading

File: feel/feelvf/ginac.cpp [hunk 10]
```diff
@@ -153,11 +153,16 @@ parse( std::string const& str, std::string const& seps, std::vector<symbol> cons
     CHECK( fsize  > 0 ) << "bad expression format";
     std::string strexpr( fields[0] );
     std::vector<std::string> strsyms;
+#if 0
```
Comment (prudhomm) [inline feel/feelvf/ginac.cpp]: remove the code here. do you think it will still serve some purpose in the future?

File: feel/feelvf/ginac.hpp [hunk 11]
```diff
@@ -504,6 +504,22 @@ expr( std::string const& s, std::vector<std::string> const& se, std::vector<Expr
     return Expr< GinacExVF<ExprT,Order> >(  GinacExVF<ExprT,Order>( g.first, g.second, s, VFmap, filename, world ) );
 }
 
+template<typename ExprT,int Order>
```
Comment (prudhomm) [inline feel/feelvf/ginac.hpp]: add documentation for function

File: feel/feelvf/ginac.hpp [hunk 12]
```diff
@@ -610,121 +626,160 @@ expr( GiNaC::ex const& f, std::vector<GiNaC::symbol> const& lsym, std::string co
...
+    for ( GiNaC::symbol const& s2 : e2.expression().symbols() )
+        exprSymb.insert( s2.get_name() );
+    for ( std::string const& s : exprSymb )
+        exprDesc += (":"+s);
+    return expr<Order>( exprDesc, filename, world, dirLibExpr );
+}
+
+template<int Order>
+inline
+Expr<GinacEx<Order> >
```
Comment (prudhomm) [inline feel/feelvf/ginac.hpp]: add Doxygen documentation

File: feel/feelvf/ginac.hpp [hunk 13]
```diff
@@ -610,121 +626,160 @@ expr( GiNaC::ex const& f, std::vector<GiNaC::symbol> const& lsym, std::string co
...
     return expr<Order>( GiNaC::diff( s.expression().expression(),{ *it },diffOrder)(0,0),
-                        s.expression().symbols(), exprDesc, filename, world );
+                        s.expression().symbols(), exprDesc, filename, world, dirLibExpr );
+}
+
+
+template<int Order,int Order1,int Order2>
+inline
+Expr<GinacEx<Order> >
+expr_mult( Expr<GinacEx<Order1>> const& e1, Expr<GinacEx<Order2>> const& e2, std::string filename="",
```
Comment (prudhomm) [inline feel/feelvf/ginac.hpp]: Add Doxygen documentation

File: testsuite/feelmodels/test.feelpp [hunk 14]
```diff
@@ -20,11 +20,11 @@
     },
     "Materials":
     {
-        "fluid":
+        "myfluid":
```
Comment (prudhomm) [inline testsuite/feelmodels/test.feelpp]: why `my`?
I would suggest `thefluid` or `fluid`

File: testsuite/feelmodels/test.feelpp [hunk 15]
```diff
@@ -33,9 +33,9 @@
             "f" : "2*g:g",
             "h" : "2*g+t:g:t"
         },
-	"copper":
+	    "mycopper":
```
Comment (prudhomm) [inline testsuite/feelmodels/test.feelpp]: as above

File: toolboxes/feel/feelmodels/electric/electric.hpp [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+namespace FeelModels
+{
+
+enum class ElectricPostProcessFieldExported
+{
+    Temperature = 0, ElectricPotential, ElectricField, Pid
+};
+
+template< typename ConvexType, typename BasisPotentialType>
+class Electric : public ModelNumerical,
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/electric/electric.hpp]: add Doxygen Documentation

File: toolboxes/feel/feelmodels/electric/electricpropertiesdescription.hpp [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+#include <feel/feelvf/cst.hpp>
+#include <feel/feelmodels/modelexpression.hpp>
+
+namespace Feel
+{
+namespace FeelModels
+{
+
+template<class SpaceType>
+class ElectricPropertiesDescription
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/electric/electricpropertiesdescription.hpp]: add Doxygen Documentation

File: toolboxes/feel/feelmodels/electric/electricpropertiesdescription.hpp [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+
+    ElectricPropertiesDescription( std::string const& prefix )
+        :
+        M_isDefinedOnWholeMesh( true ),
+        M_electricConductivityDefaultValue( doption(_name="electric-conductivity",_prefix=prefix) )
+        {}
+
+    ElectricPropertiesDescription( ElectricPropertiesDescription const& ) = default;
+
+    void updateForUse( mesh_ptrtype const& mesh , ModelMaterials const& mats, std::vector<WorldComm> const& worldsComm )
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/electric/electricpropertiesdescription.hpp]: big function in cpp file instead

File: toolboxes/feel/feelmodels/electric/electricpropertiesdescription.hpp [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+            return M_electricConductivityByMaterial.find( matName ) != M_electricConductivityByMaterial.end();
+        }
+    ModelExpressionScalar const& electricConductivity( std::string const& matName ) const
+        {
+            CHECK( this->hasElectricConductivity( matName ) ) << "material name not registered : " << matName;
+            return M_electricConductivityByMaterial.find( matName )->second;
+        }
+
+    boost::shared_ptr<std::ostringstream>
+    getInfoMaterialParameters() const
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/electric/electricpropertiesdescription.hpp]: cpp file instead

File: toolboxes/feel/feelmodels/electric/electricpropertiesdescription.hpp [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+                *ostr << "\n     -- [" << matName << "] electric conductivity : ";
+                if ( this->electricConductivity(matName).isConstant() )
+                    *ostr << this->electricConductivity(matName).value();
+                else
+                    *ostr << str( this->electricConductivity(matName).expr().expression() );
+            }
+            return ostr;
+        }
+
+    bool hasElectricConductivityDependingOnSymbol( std::string const& symbolStr ) const
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/electric/electricpropertiesdescription.hpp]: big function in cpp file instead

File: toolboxes/feel/feelmodels/fluid/densityviscositymodel.hpp [hunk 21]
```diff
@@ -59,43 +59,47 @@ public :
         }
     DensityViscosityModel( DensityViscosityModel const& app  ) = default;
 
-    void updateForUse( mesh_ptrtype const& mesh , ModelMaterials const& mat, std::vector<WorldComm> const& worldsComm, bool useExtendedDofTable )
+    void updateForUse( mesh_ptrtype const& mesh , ModelMaterials const& mats, bool useExtendedDofTable )
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/fluid/densityviscositymodel.hpp]: big function in cpp file instead

File: toolboxes/feel/feelmodels/fluid/densityviscositymodel.hpp [hunk 22]
```diff
@@ -59,43 +59,47 @@ public :
...
+
+                if ( mat.hasPropertyExprScalar("rho") )
+                    this->setDensity( mat.propertyExprScalar("rho"),matmarker );
+                else
+                    this->setCstDensity( mat.propertyConstant("rho"),matmarker );
+            }
         }
     }
-    void updateForUse( space_ptrtype const& space, ModelMaterials const& mat )
+    void updateForUse( space_ptrtype const& space, ModelMaterials const& mats )
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/fluid/densityviscositymodel.hpp]: big function in cpp file instead

File: toolboxes/feel/feelmodels/fluid/dynamicviscositymodel.hpp [hunk 23]
```diff
@@ -84,7 +84,7 @@ public :
         }
     DynamicViscosityModel( DynamicViscosityModel const& app  ) = default;
 
-    void updateForUse( mesh_ptrtype const& mesh , ModelMaterials const& mat, std::vector<WorldComm> const& worldsComm, bool useExtendedDofTable )
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/fluid/dynamicviscositymodel.hpp]: big function in cpp file

File: toolboxes/feel/feelmodels/fluid/dynamicviscositymodel.hpp [hunk 24]
```diff
@@ -94,61 +94,80 @@ public :
...
+            {
+                if ( mat.hasPropertyExprScalar("mu") )
+                    this->setDynamicViscosity( mat.propertyExprScalar("mu"),matmarker );
+                else
+                    this->setCstDynamicViscosity( mat.propertyConstant("mu"), matmarker );
+            }
         }
     }
-    void updateForUse( space_ptrtype const& space, ModelMaterials const& mat )
+    void updateForUse( space_ptrtype const& space, ModelMaterials const& mats )
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/fluid/dynamicviscositymodel.hpp]: as above

File: toolboxes/feel/feelmodels/heat/heat.hpp [hunk 25]
```diff
@@ -52,16 +50,16 @@ namespace Feel
 namespace FeelModels
 {
 
-
 template< typename ConvexType, typename BasisTemperatureType>
-class ThermoDynamicsBase : public ModelNumerical,
-                           public MarkerManagementDirichletBC,
-                           public MarkerManagementNeumannBC,
-                           public MarkerManagementRobinBC
+class Heat : public ModelNumerical,
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/heat/heat.hpp]: add Doxygen Documentation

File: toolboxes/feel/feelmodels/heatfluid/heatfluid.hpp [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+#include <feel/feelmodels/fluid/fluidmechanics.hpp>
+
+
+namespace Feel
+{
+namespace FeelModels
+{
+
+template< typename HeatType, typename FluidType>
+class HeatFluid : public ModelNumerical,
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/heatfluid/heatfluid.hpp]: add Doxygen Documentation

File: toolboxes/feel/feelmodels/heatfluid/heatfluid.hpp [hunk 27]
```diff
@@ -0,0 +1,160 @@
...
+
+private :
+    void loadParameterFromOptionsVm();
+    void initMesh();
+    void initPostProcess();
+public :
+    // update for use
+    void init( bool buildModelAlgebraicFactory = true );
+
+    BlocksBaseGraphCSR buildBlockMatrixGraph() const;
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/heatfluid/heatfluid.hpp]: basic documentation should be available for a function like this (use eg //!)

File: toolboxes/feel/feelmodels/levelset/levelset.cpp [hunk 28]
```diff
@@ -89,9 +89,9 @@ LEVELSET_CLASS_TEMPLATE_TYPE::New(
         std::string const& prefix,
         WorldComm const& worldComm,
         std::string const& subPrefix,
-        std::string const& rootRepository )
+        ModelBaseRepository const& modelRep )
 {
-    self_ptrtype new_ls( new self_type(prefix, worldComm, subPrefix, rootRepository) );
+    self_ptrtype new_ls( new self_type(prefix, worldComm, subPrefix, modelRep ) );
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/levelset/levelset.cpp]: use `return make_shared<self_type>(...)`

File: feel/feelcore/environment.hpp (hunk 1 above)
Comment (vincentchabannes) [reply feel/feelcore/environment.hpp]: it's for store expression (or other data) which not depends on number of proc (it was implemented in toolboxes core but it easier to manage this in the Environment.

File: toolboxes/feel/feelmodels/fluid/dynamicviscositymodel.hpp (hunk 23 above)
Comment (prudhomm) [reply toolboxes/feel/feelmodels/fluid/dynamicviscositymodel.hpp]: This generate code bloat
